src/openloco/windows/townwnd.cpp

```diff
@@ -1,31 +1,863 @@
+                self->widgets[widx::demolish_town].type = widget_type::wt_9;
+            }
+            else
+            {
+                self->widgets[widx::expand_town].type = widget_type::none;
+                self->widgets[widx::demolish_town].type = widget_type::none;
+                self->widgets[widx::viewport].right += 22;
+            }
+
+            self->widgets[widx::centre_on_viewport].right = self->widgets[widx::viewport].right - 1;
+            self->widgets[widx::centre_on_viewport].bottom = self->widgets[widx::viewport].bottom - 1;
+            self->widgets[widx::centre_on_viewport].left = self->widgets[widx::viewport].right - 24;
+            self->widgets[widx::centre_on_viewport].top = self->widgets[widx::viewport].bottom - 24;
+
+            common::repositionTabs(self);
+        }
+
+        // 0x00498FFE
+        static void draw(window* self, gfx::drawpixelinfo_t* dpi)
+        {
+            self->draw(dpi);
+            common::drawTabs(self, dpi);
+            self->drawViewports(dpi);
+            widget::drawViewportCentreButton(dpi, self, (widget_index)widx::centre_on_viewport);
+
+            static string_id townSizeLabelIds[] = {
+                string_ids::town_size_hamlet,
+                string_ids::town_size_village,
+                string_ids::town_size_town,
+                string_ids::town_size_city,
+                string_ids::town_size_metropolis,
+            };
+
+            auto town = townmgr::get(self->number);
+            commonFormatArgs[0] = townSizeLabelIds[(uint8_t)town->size];
+            *(int32_t*)&commonFormatArgs[1] = town->population;
+
+            const uint16_t xPos = self->x + self->widgets[widx::unk_8].left - 1;
+            const uint16_t yPos = self->y + self->widgets[widx::unk_8].top - 1;
+            const uint16_t width = self->widgets[widx::unk_8].width() - 1;
+            gfx::draw_string_494BBF(*dpi, xPos, yPos, width, colour::black, string_ids::status_town_population, &*commonFormatArgs);
+        }
+
+        // 0x00499079
+        static void on_mouse_up(window* self, widget_index widgetIndex)
+        {
+            switch (widgetIndex)
+            {
+                case common::widx::caption:
+                    common::renameTownPrompt(self, widgetIndex);
+                    break;
+
+                case common::widx::close_button:
+                    WindowManager::close(self);
+                    break;
+
+                case common::widx::tab_town:
+                case common::widx::tab_population:
+                case common::widx::tab_company_ratings:
+                    common::switchTab(self, widgetIndex);
+                    break;
+
+                case widx::centre_on_viewport:
+                {
+                    if (self->viewports[0] == nullptr || self->saved_view.isEmpty())
+                        break;
+
+                    registers regs;
+                    if (self->saved_view.hasUnkFlag())
```

Comment:
Will never happen

Reply:
You'll have to elaborate here. There's an explicit check for this at `0x0049934C`, and normally you're rather strict about these things…

Reply:
The original code has dead branches, since a town window will never focus on a `thing`. `SavedView` is the same as `edx` and `ecx` in `viewportmgr::create(registers regs, int index)`. It's basically used as a hash, in case the postion of the town (?) changes.